### ISSUE TITLE
testClone2 failure with g++ 5.4.0 (Ubuntu 16.04)

### ISSUE THREAD

**Issue by rolanddenis** [DGtal-team/DGtal#1203]
`testClone2` compiles fine but fails at runtime (see https://gist.github.com/rolanddenis/842ac840e76c6d70a75c787ce14d0cfa ).

Everything works fine with g++ 5.2.1 (Ubuntu 15.10).

**Comment by JacquesOlivierLachaud** [comment]
Hum, only idea at the moment is that g++ 5.4 optimized out too much. 
I'll have a look in September. 



De: "rolanddenis" <notifications@github.com> 
À: "DGtal-team/DGtal" <DGtal@noreply.github.com> 
Envoyé: Mardi 9 Août 2016 10:41:23 
Objet: [DGtal-team/DGtal] testClone2 failure with g++ 5.4.0 (Ubuntu 16.04) (#1203) 



testClone2 compiles fine but fails at runtime (see https://gist.github.com/rolanddenis/842ac840e76c6d70a75c787ce14d0cfa ). 

Everything works fine with g++ 5.2.1 (Ubuntu 15.10).

**Comment by dcoeurjo** [comment]
it is also reported by hombrew-science: https://github.com/Homebrew/homebrew-science/pull/3787#issuecomment-241789853

**Comment by ilovezfs** [comment]
I tried with `ENV.O0` and `ENV["HOMEBREW_OPTFLAGS"] = "-march=#{Hardware.oldest_cpu}"` (core2), and still
```
Start  24: testClone2
24/183 Test  #24: testClone2 .................................***Failed    0.00 sec
```

**Comment by ilovezfs** [comment]
For Homebrew, I think I will add
```
  # GCC 4 works, and according to upstream issue, GCC <= 5.2.1 may also be fine
  ["5", "6"].each do |n|
    fails_with :gcc => n do
      cause "testClone2 fails: https://github.com/DGtal-team/DGtal/issues/1203"
    end
  end
```
until this is sorted out so that the 0.9.2 release doesn't need to be deferred any longer. Does that make sense as a plan?

**Comment by dcoeurjo** [comment]
Good idea. Let's do it like that for the homebrew formula while we are investigating

**Comment by rolanddenis** [comment]
At first glance, it seems related to this line in `testClone2.cpp`:
```C++
inline CloneToCountedMember( Clone<DummyTbl> a1 )
  : myDummyTbl( new DummyTbl( a1 ) ) {}
```
( see https://github.com/DGtal-team/DGtal/blob/master/tests/base/testClone2.cpp#L244 )

`new DummyTbl( a1 )` calls `Clone<DummyTbl>::operator T()` that involves a first copy construction (see https://github.com/DGtal-team/DGtal/blob/master/src/DGtal/base/Clone.h#L365 and https://gist.github.com/rolanddenis/842ac840e76c6d70a75c787ce14d0cfa#file-gistfile1-txt-L27) and the new instance is then moved during the heap allocation (https://gist.github.com/rolanddenis/842ac840e76c6d70a75c787ce14d0cfa#file-gistfile1-txt-L30).

From there, the behaviour of g++ 5.4 doesn't seem wrong but I don't understand why it does not use Return Value Optimization like previous versions ... Actually, this test also failed with previous gcc versions when adding `-fno-elide-constructors` option to disable copy/move elisions.

Firstly, it is possible to avoid this bug by replacing
```C++
  : myDummyTbl( new DummyTbl( a1 ) ) {}
```
with
```C++
  : myDummyTbl( &a1 ) {}
```
but considering the last remark about the result of `-fno-elide-constructors`, it is probably safer to use separates counters for move constructions/destructions.

**Comment by dcoeurjo** [comment]
thanks a lot @rolanddenis for this in-depth analysis. @JacquesOlivierLachaud what do you think?

**Comment by dcoeurjo** [comment]
@JacquesOlivierLachaud could you please have a look to this issue?

**Comment by JacquesOlivierLachaud** [comment]
Should be corrected by PR #1382
